datamapper/changeset/changeset.py

```diff
@@ -82,12 +72,22 @@ def get_field(self, field: str, default: Any = None) -> Optional[Any]:
             field, default
         )
 
-    def validate_required(self, fields: List) -> Changeset:
+    def has_field(self, field: str) -> bool:
+        """
+        Returns `True` if the field is present in the changes or data.
+        """
+        return field in self.changes or field in self._wrapped_data.attributes
+
+    def validate_required(
+        self, fields: List, message: str = "Missing data for required field."
+    ) -> Changeset:
         """
         Validate that `fields` are present in the changeset, either as changes or as data.
         """
-        for f in fields:
-            self.required.add(f)
+        for field in fields:
+            if not self.has_field(field):
```

Comment:
In Ecto, I think the required validation actually checks for `None` and empty strings:

> If the value of a field is nil or a string made only of whitespace, the changeset is marked as invalid, the field is removed from the changeset's changes, and an error is added. An error won't be added if the field already has an error.

Also, the error message in Ecto is "can't be blank"

Reply:
Ah, good call. I forgot that `cast` takes an `empty_values` option which defaults to `""`.